llvm/lib/Target/AMDGPU/AMDGPUOclcReflect.cpp

```diff
@@ -0,0 +1,92 @@
+//===- AMDGPUOclcReflect.cpp ----------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass searches for occurences of the AMDGPU_OCLC_REFLECT function, and
+// replaces the calls with some val dependent on the operand of the func. This
+// can be used to reflect across different implementations of functions at
+// compile time based on a compiler flag or some other means. This pass
+// currently supports use cases:
+//
+// 1. Choose a safe or unsafe version of atomic_xor at compile time, which can
+//    be chosen at compile time by setting the flag
+//    --amdgpu-oclc-unsafe-int-atomics=true.
+//
+// This pass is similar to the NVPTX pass NVVMReflect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "AMDGPU.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/CommandLine.h"
+
+using namespace llvm;
+
+#define AMDGPU_OCLC_REFLECT "__oclc_amdgpu_reflect"
+
+static cl::opt<bool>
+    AMDGPUReflectEnabled("amdgpu-oclc-reflect-enable", cl::init(true),
+                         cl::Hidden,
+                         cl::desc("AMDGPU reflection, enabled by default"));
+static cl::opt<bool> AMDGPUUnsafeIntAtomicsEnable(
+    "amdgpu-oclc-unsafe-int-atomics", cl::init(false), cl::Hidden,
+    cl::desc("Should unsafe int atomics be chosen. Disabled by default."));
+
+PreservedAnalyses AMDGPUOclcReflectPass::run(Function &F,
+                                             FunctionAnalysisManager &AM) {
+  if (!AMDGPUReflectEnabled)
+    return PreservedAnalyses::all();
+
+  if (F.getName() == AMDGPU_OCLC_REFLECT) {
+    assert(F.isDeclaration() &&
+           "__oclc_amdgpu_reflect function should not have a body");
+    return PreservedAnalyses::all();
+  }
+
+  SmallVector<CallInst *, 4> ToRemove;
+
+  for (Instruction &I : instructions(F)) {
+    CallInst *Call = dyn_cast<CallInst>(&I);
```

Comment:
```suggestion
    auto *Call = dyn_cast<CallInst>(&I);
```

Pretty much nit, can be ignored

Reply:
Change made